src/TuxState.js

```diff
@@ -0,0 +1,650 @@
+    invariantArgCheck(propsToShift);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
+    //build new state object
+    var newState = buildNewState(currentState, propsToShift, function(newState, key, currentStateAtKey, newPropsAtKey) {
+      //check that current state at this key is an array
+      invariantArrayCheck(currentStateAtKey);
+      //shift first value off the corresponding array in the current state
+      currentStateAtKey.shift();
+      newState[key] = currentStateAtKey;
+    });
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+  //spliceState FUNCTION: splice the values of the deepest keys in the current state with the array of arguments in the corresponding keys in the passed in object. Throws an error if keys don't match or if the deepest key in current state is not an array
+  //@param propsToSplice OBJECT: required object argument with deepest keys being an array of arguments to invoke splice with
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  spliceState: function(propsToSplice, callback) {
+    //throw error if propsToSplice is undefined or not an object
+    invariantArgCheck(propsToSplice);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
+    //build new state object
+    var newState = buildNewState(currentState, propsToSplice, function(newState, key, currentStateAtKey, newPropsAtKey) {
+      //check that current state at this key is an array
+      invariantArrayCheck(currentStateAtKey);
+      //check that newPropsAtKey is an array of arguments to be passed into the splice invocation
+      invariantArrayCheck(newPropsAtKey);
+      //invoke splice on currentStateAtKey with the array of arguments in newPropsAtKey
+      Array.prototype.splice.apply(currentStateAtKey, newPropsAtKey);
+      newState[key] = currentStateAtKey;
+    });
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+  //concatToEndOfState FUNCTION: concat the arrays of values in the deepest keys on the passed in object to the end of the corresponding arrays in the deepest keys in the current state. Throws an error if keys don't match or if the deepest key in current state is not an array
+  //@param propsToShift OBJECT: required object argument with deepest keys being a boolean
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  concatToEndOfState: function(propsToConcat, callback) {
+    //throw error if propsToConcat is undefined or not an object
+    invariantArgCheck(propsToConcat);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
```

Comment:
I don't think we need to invoke `assign` here because we are already making a shallow copy of state at the first line in `buildNewState`.